### DISCUSSION TITLE
Where to maintain compose/stack files for multi-repo project?

### DISCUSSION THREAD

**Discussion by mistresseve666** [BretFisher/ama#130] [Q&A]
Hello,

For projects that separate the frontend and the various backend micro-services into separate repos, where would it be best to keep the compose/stack files? For example, in the classic dogs vs cats voting app if you have the following repos:

- frontend
- api gateway
- vote
- results
- worker

Would it be best to have a single "deploy" repo that maintains a single stack file and supporting config/secrets? Or separate the stack file out amongst the 5 various services?

The former seems simpler to deploy as you'd be able to deploy a single stack and easily define the links between the various services. However, you then need to manually update the repo once a stable release of each service is made.

With the former each service is self-contained in its own repo, can take advantage of CI to automatically deploy the new service on tag, etc. However, the deployment is made more complex by needing to define links somehow between the services. For example, vote talks to redis which is also accessed by worker, etc.

Thanks,
-robodude666

**Comment by guillaumemaka** [comment]
I'm also interested of your approach, I have a similar case and I try to figure out how to organise my repositories to add CI and CD (I'll use drone.io).

- Web Api
- Admin Dashboard Frontend
- Public frontend
- worker with NATS

I made a repo with the 3 aboves projects (with a Dockerfile in each) as submodules + a docker-stack.yml to make development more easier.

Now I try to figure out how to do the deploy update  when I build a new image

For now I have 2 approaches in my mind:

**Approach 1:** 

1. Build the new image
2. tag the new built image with the commit/tag then map that version to latest
3. fetch the repo with all the project as submodules + docker-stack.yml
4. exec docker stack deploy -c docker-stack.yml <stack_name>

I assume that docker will figure out that the latest tag digest of each image has changed and update the stack to use the new image

**Approach 2:** 

1. Somehow hardcode the service name and stack name in my CI/CD configuration
2. Build the new image
3. tag the new built image with the commit/tag then map that version to latest
4. exec docker service update  --image <the_new_built_image:tag> <stack_name>_<service_name>

So it would be great to have your opinion about that.

Thanks for the great content you made so far. (Hope to see a K8 course soon)

**Comment by mistresseve666** [comment]
@guillaumemaka you'd want to avoid using the tag `latest` because of its ambiguity. You never know what latest is. You'd want to reference `v2.0.0` explicitly, etc.

The submodule approach is interesting, though I'm not a huge fan of them. To my understanding you can deploy a stack of the same name from different stack files as long as each one has different service names. This would allow a single stack to exist for your application but be deployed from CI of different repos.

When it comes to Docker Swarm it seems most (serious) people end up using Docker Enterprise edition. Based on my understanding, Docker EE is capable of updating a service running on the UCP automatically if a new image is pushed to DTR.

Also I came across [Docker App](https://github.com/docker/app) which allows for metadata+compose+variables to be pushed as a single package to the registry. Currently it is a stand alone plugin but should be included in docker 19.03.

After more thought on this topic I think the "better" solution is to keep the stack files and CI/CD pipelines in the same repo as the service itself. This will mean for the example app above you'd have 5 separate stack files being deployed with their own services.

@BretFisher what are your thoughts?

**Comment by BretFisher** [comment]
When you're on a project with 1. more than one person and 2. more then a few repos, I think it makes sense to put the compose files in their own repo. Think of it as a GitOps repo.

In the upcoming 19.03 docker release `docker app` replaces the concept tool `docker-app`. It'll help you with templating and shipping stack files in images just like you ship code.

You don't want to change your stacks based on your git repo organization.  A stack file should be designed to include everything your app needs (assuming those things are unique for that stack).

If you're deploying apps with stacks, you don't want to use `service update` later. Set your stack file up with templating features like environment variables so your CI/CD can plug in image tags when you re-deploy the stack based on image builds. Each image built would use something like git commit ID or CI job ID as the image tag, and then pass it to the `stack deploy` as environment variables.

Never reuse tags on servers. Only use `latest` for branch-name based tags for developer convenience when you just need to quickly run a thing locally.

Did that answer all the questions?

**Comment by mistresseve666** [comment]
> When you're on a project with 1. more than one person and 2. more then a few repos, I think it makes sense to put the compose files in their own repo. Think of it as a GitOps repo.

Bret, I want to make sure I understand this statement. By "in their own repo" do you mean a single repo, correct? Or by "own" you mean the vote app files go in the vote repo, worker compose files go into the worker repo, etc.

**Comment by BretFisher** [comment]
I was specifically talking about the compose file, not the apps themselves. App design in terms of git repos is quite diverse. Dockerfiles go with the code. Compose files are at a higher level of abstraction, so they go more with DevOps stuff in a separate repo.  

Generally the smaller the team and apps, the less things need to be broken out into separate git repos. There is no one-model-fits-all approach. 

The one thing to stress is that compose/stack/kubernetes YAML should be designed based on the app architecture, not based on "parts of the YAML go in different git repos".

**Comment by mistresseve666** [comment]
Thanks! Do you have any articles/videos you'd recommend that discuss maintaining a "DevOps" repo and how it's used during normal development? Are there any example projects/repos that show this in practice?

With either approach I'm not certain how the dev process would go. If you wanted to debug the whole app you'd need to manually start each individual service?

In your [Live Show Ep42](https://youtu.be/qf_w79H2JhA?t=3425) you talked about having automation in place before splitting services apart. Are you recommending using a DevOps repo here as it's simpler for most people/orgs to implement, but you'd ultimate recommend to strive for separating the compose files with automation?

**Comment by BretFisher** [comment]
Since there is no one-size-fits-all here, it's hard to give you specifics because I know zero about your needs. Using the dogvs.cat you mentioned, with only a few developers, you might have:
- one git repo per app
- one git repo for DevOps related yaml, scripts, etc.

Your compose would put a single compose file in the DevOps repo, where `docker-compose up` would spin up the full solution locally. It points to the `context` of each service under their local path, assuming all repos are cloned into a single root folder, so:

```
~/code/dogvscat
~/code/dogvscat/dogvscat-devops
~/code/dogvscat/dogvscat-worker
etc.
```

You might have a similar stack YAML for Swarm that would spin up the full solution in a single stack on a single Swarm. That plus your infrastructure scripts (Terraform, Ansible, etc.) would all be in DevOps until you had a good reason to split something out to a sperate repo (maybe you don't want all devs to have your Ansible scripts 🤷‍♂️)

When a developer is getting started, they:
1. install docker locally
2. clone all the repos under a dogvscat directory
3. go in the DevOps repo and do a `docker-compose up`.

**Comment by mistresseve666** [comment]
Would maintaining the application as a single stack file be desired?

If each microservice is separated into its own stack it would enable being able to deploy changes to that service without touching the other services. It would also increase security and prevent microservices from communicating with one another (or their databases) unless permitted by the network rules.

**Comment by BretFisher** [comment]
yes in the above example it should all be one stack. You can make as many networks as you need in the single stack file to control network access. Swarm knows when something changes on each stack deploy, so it only changes what is needed.
